fix: manage CURRENT_TIMESTAMP(6) correctly

law108000 opened the pull request:
### Description of change
just fix the issue with CURRENT_TIMESTAMP(6) being used, can "DEFAULT" be used?

while I use the code like this.
```javascript
await this.parentRepository
        .createQueryBuilder()
        .relation("relation")
        .of(parentRecord)
        .add(childRecord);
```
The generated SQL is always hardcoded to set the updated_date to be CURRENT_TIMESTAMP.
```sql
UPDATE `child` SET `parentId` = ?, `updated_date` = CURRENT_TIMESTAMP WHERE `id` IN (?)
```
After investigating all related source code, I found the below TODO in UpdateQueryBuilder.ts about 2 years ago:
```
todo: fix issue with CURRENT_TIMESTAMP(6) being used, can "DEFAULT" be used?!
```




### Pull-Request Checklist

<!--
  Please make sure to review and check all of the following.

  If an item is not applicable, you can add "N/A" to the end.
-->

- [x] Code is up-to-date with the `master` branch
- [ ] `npm run format` to apply prettier formatting
- [ ] `npm run test` passes with this change
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

<!--
  🎉 Thank you for contributing and making TypeORM even better!
-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced comprehensive test suites to verify correct handling of timestamp and datetime precision (0 and 6) in entity updates and relations across multiple database drivers (MySQL, MariaDB, Postgres, Oracle, SQL Server, SQLite).
  * Added new entity classes and base classes for each supported precision and database, supporting automated timestamp updates with appropriate precision.

* **Tests**
  * Added extensive tests to ensure update operations use the correct timestamp/datetime precision and SQL functions for each database.
  * Verified schema creation, SQL generation, and update behavior for both precision 0 and 6 across all supported drivers.

* **Refactor**
  * Improved maintainability by centralizing column assignment logic for update queries, reducing code duplication.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

law108000 committed:
fix: issue with CURRENT_TIMESTAMP(6) being used, can "DEFAULT" be used

law108000 commented:
> Not sure what issue we are trying to fix here.
> 
> Default is the column default value (in the column definition in the create table statement), the one that is used when a row is inserted, not the default update value.

Oh, sorry for the incomplete story that confused you.
In my use case, I want high-precision create and update dates, so i config to use `CURRENT_TIMESTAMP(6)` instead of the defaulted `CURRENT_TIMESTAMP`.

my configuration is like the below:
```typescript
class MyEntity {
  @CreateDateColumn({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP(6)', transformer: new DateTransformer() })
  @IsOptional()
  @IsISO8601()
  create_date?: Date;

  @UpdateDateColumn({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP(6)', transformer: new DateTransformer() })
  @IsOptional()
  @IsISO8601()
  updated_date?: Date;
  
  // others attrs...
}
```

I thought the TypeORM would be able to handle it well, but it failed on my jest test code.
for example, the correct updated_date should be `2023-05-23 02:53:44.143000`, but typeorm always stored it as `2023-05-23 02:53:44.000000`.

after digging into the TypeORM source code, I found it was always hardcoded to `" = CURRENT_TIMESTAMP"` no matter what I set on `CreateDateColumn`, and there left a TODO comment.

alumni commented:
Okay, I get it now :) I think it makes sense.

I'm a bit surprised that `CURRENT_TIMESTAMP` was hard-coded in that place. Normally, this is driver dependent - `Driver.mappedDataTypes.updateDateDefault` has the function the gets the current time. So `updateDateColumn.default` should always be set to this value:
https://github.com/typeorm/typeorm/blob/27b4207c48a391c7bf7013b251f756f48861c496/src/metadata/ColumnMetadata.ts#L504-L521

Since the default is always set, I think we can always use it. What do you think?

law108000 committed:
Merge branch 'typeorm:master' into master

law108000 committed:
fix: use default from column metadata config instead of hardcode

law108000 commented:
> Okay, I get it now :) I think it makes sense.
> 
> I'm a bit surprised that `CURRENT_TIMESTAMP` was hard-coded in that place. Normally, this is driver dependent - `Driver.mappedDataTypes.updateDateDefault` has the function the gets the current time. So `updateDateColumn.default` should always be set to this value:
> 
> https://github.com/typeorm/typeorm/blob/27b4207c48a391c7bf7013b251f756f48861c496/src/metadata/ColumnMetadata.ts#L504-L521
> 
> Since the default is always set, I think we can always use it. What do you think?

Awesome! Thanks for suggesting the default config in ColumnMetadata! It really ties everything together perfectly!
I just made a change on the pull request. please take a look if it is okay. Thank you.

alumni commented on src/query-builder/UpdateQueryBuilder.ts:
```diff
@@ -642,11 +642,16 @@ export class UpdateQueryBuilder<Entity extends ObjectLiteral>
+                    }
+                    updateColumnAndValues.push(this.escape(updateDateColumn.databaseName) +
+                        " = " +
+                        defaultValue);
```
I would somehow do this:
```ts
valuesSetNormalized[updateDateColumn.databaseName] = updateDateColumn.default
```
and let the `else` block handle the whole thing 🤷🏻‍♂️

Why? `defaultValue` needs to be handled properly, e.g. if it's not a function it needs to be escape. Currently this is already implemented in the `else` block.

law108000 committed:
fix: handle updateDateColumn via valuesSetNormalized

law108000 committed:
Merge branch 'typeorm:master' into master

law108000 committed:
refactor: update value processing in UpdateQueryBuilder

Extracts value processing logic into a new protected createColumnAssignment method to eliminate code duplication and improve maintainability. The update logic now distinguishes between property paths and database names, ensuring correct handling of spatial types, SQL expressions, and driver-specific behaviors.

law108000 committed:
refactor: update column handling in UpdateQueryBuilder

Simplifies the logic for processing update columns by removing the separation of property paths and database names. Ensures the update date column is set only if not already updated, and refactors how default values are applied for the update date column.

alumni commented:
Nice, @law108000. A test would be great to prevent this feature from breaking in the future.

law108000 committed:
refactor: update date column assignment logic

Simplifies the assignment of the update date column by using the createColumnAssignment method instead of manually handling default values and escaping. This improves code readability and maintainability.

law108000 committed:
test: Add tests for timestamp precision in UpdateDateColumn

Introduces comprehensive tests for handling CURRENT_TIMESTAMP and CURRENT_TIMESTAMP(6) in UpdateDateColumn, covering both precision 0 and precision 6 scenarios. Tests verify correct SQL generation, table creation, and timestamp updates for entity relations and direct updates, ensuring proper handling across supported databases.

law108000 committed:
test: Add SQL Server precision tests for issue #11258

Introduces SQL Server-specific test entities and a dedicated test suite to verify datetime2 precision handling with GETDATE() and SYSDATETIME() for UpdateDateColumn. Updates the main issue-11258 test to clarify MySQL/MariaDB focus, add helper functions, and skip SQL Server for incompatible tests. Ensures both schema and SQL queries use the correct timestamp precision for relation and entity updates.

alumni commented on test/github-issues/11258/issue-11258.ts:
```diff
@@ -0,0 +1,796 @@
+                __dirname + "/entity_precision_6/Child{.js,.ts}",
+                __dirname + "/entity_precision_6/Base{.js,.ts}",
+            ],
+            enabledDrivers: ["mysql", "mariadb", "oracle", "postgres"],
```
`sap` can be added if you do any adjustments

alumni commented on test/github-issues/11258/issue-11258-sqlserver.ts:
```diff
@@ -0,0 +1,594 @@
+                    },
+                )
+
+                await new Promise((res) => setTimeout(res, 10))
```
```ts
import { scheduler } from 'node:timers/promises';
await scheduler.wait(10);
```

law108000 committed:
Update test/github-issues/11258/issue-11258.ts

Co-authored-by: Lucian Mocanu <alumni@users.noreply.github.com>

law108000 committed:
Update test/github-issues/11258/issue-11258.ts

Co-authored-by: Lucian Mocanu <alumni@users.noreply.github.com>

law108000 committed:
Update test/github-issues/11258/issue-11258.ts

Co-authored-by: Lucian Mocanu <alumni@users.noreply.github.com>

law108000 committed:
refactor: SQL Server datetime2 precision tests

Replaces setTimeout-based delays with scheduler.wait from node:timers/promises for improved readability and reliability. Updates comments and entity imports for clarity, removes redundant millisecond assertions, and enhances precision 6 timestamp checks to ensure microsecond precision is preserved.

law108000 committed:
test: Add SQLite precision tests for UpdateDateColumn

Introduces new test entities and test cases for SQLite to verify correct handling of datetime precision (0 and 6) with UpdateDateColumn. Also refactors entity creation helpers in SQL Server and MySQL/MariaDB tests for consistency and type safety.

alumni commented:
The SAP tests fail because in HANA the timestamp precision must be between 1 and 7 (there's a different data type for precision 0), so creating the connection fails during schema setup. Let's say we can skip the SAP tests for now.

I think the other reviewers might comment on the test code quality.

alumni committed:
test: quick lint+style fixes

law108000 committed:
Update issue-11258-mssql.ts

emmettprexus commented:
Any update on this? I just ran into the same challenge of getting my updatedAt to ms precision and finally ended up here...

gioboa committed:
Merge branch 'master' into master

gioboa committed:
Merge branch 'master' into master

gioboa committed:
Merge branch 'master' into master

gioboa committed:
Merge branch 'master' into master

gioboa committed:
test: rename files 🍍

gioboa committed:
Merge branch 'master' into master

gioboa reviewed (changes requested):
is this PR still valid? can you solve the conflicts please?

Commit:
Merge remote-tracking branch 'origin/master' into work-11258

# Conflicts:
#	src/query-builder/UpdateQueryBuilder.ts
